arrow-fx-android/src/main/kotlin/arrow/integrations/android/extensions.kt

```diff
@@ -0,0 +1,43 @@
+package arrow.integrations.android
+
+import androidx.lifecycle.Lifecycle.Event
+import androidx.lifecycle.Lifecycle.State
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.LifecycleOwner
+import arrow.fx.IO
+import arrow.fx.IOOf
+import arrow.fx.IOResult
+import arrow.fx.fix
+
+/**
+ * Unsafely run an [IO] and receive the values in a callback [cb] while participating in structured concurrency.
+ * Equivalent of [IO.unsafeRunAsyncCancellableEither] but with its cancellation token wired to [LifecycleOwner].
+ *
+ * @see [IO.unsafeRunAsyncCancellableEither] for a version that returns the cancellation token instead.
+ */
+fun <E, A> LifecycleOwner.unsafeRunIO(io: IOOf<E, A>, cb: (IOResult<E, A>) -> Unit): Unit =
+  io.unsafeRunScoped(this, cb)
+
+/**
+ * Unsafely run an [IO] and receive the values in a callback [cb] while participating in structured concurrency.
+ * Equivalent of [IO.unsafeRunAsyncCancellableEither] but with its cancellation token wired to [LifecycleOwner].
+ *
+ * @see [IO.unsafeRunAsyncCancellableEither] for a version that returns the cancellation token instead.
+ */
+fun <E, A> IOOf<E, A>.unsafeRunScoped(
+  owner: LifecycleOwner,
+  cb: (IOResult<E, A>) -> Unit
+) {
+  if (owner.lifecycle.currentState.isAtLeast(State.CREATED)) {
```

Comment:
Why is this initial state required?

Reply:
Otherwise we potentially might launch an IO after the onDestroy phase and therefore might never be cancelled

Reply:
Doesn't it feel a bit confusing from the user perspective? If the condition is not matched the dev doesn't get any error but the job is not run.

If the condition is not matched we could throw, if we wanna be strict (this is already an unsafe run), or alternatively, we could install a new lifecycle observer in that case, so it runs the job automatically if the lifecycle reaches de CREATED state afterwards. (Note this could get called earlier in the lifecycle).

I've not dug into the implementation yet, but I'd say something like that is what the [Android lifecycle aware coroutines](https://developer.android.com/topic/libraries/architecture/coroutines#suspend) do under the hood.

Reply:
> (Note this could get called earlier in the lifecycle)

I don't think so, at least from what I understood here https://developer.android.com/reference/androidx/lifecycle/Lifecycle.State#summary the CREATED state is the minimum active state you can have

> or alternatively, we could install a new lifecycle observer in that case, so it runs the job automatically if the lifecycle reaches de CREATED state afterwards. 

As far as I'm concerned that wouldn't work because of this (taken from the `launchWhenXXX` fun)

![image](https://user-images.githubusercontent.com/3294640/82827722-de0bc480-9eaf-11ea-9f35-ad774a41c49d.png)

(where `handleDestroy` is simply cancelling the parent and finishing)

> I've not dug into the implementation yet, but I'd say something like that is what the Android lifecycle aware coroutines do under the hood with those launchWhenXXXX methods.

Note that we're in the `unsafeRunScoped` function which is different from the `launchWhenXXX`, the latter can be easily implementing by simply flatmapping an `ioOfEvent` to the IO we want to start at that state:

```
IO.async { cb ->	
    owner.lifecycle.addObserver(object : LifecycleEventObserver {	
      override fun onStateChanged(source: LifecycleOwner, event: Event) {	
        if (event == eventTarget) {	
          source.lifecycle.removeObserver(this)	
          cb(IOResult.Success(Unit))	
        }	
      }	
    })	
  }
```